### PR TITLE
rgw/lua: post request context does not work

### PR DESCRIPTION
**Pull request by yuvalif**
regression introduced by: 644c99826d73174e6609aa24b7297443358488b1

Signed-off-by: Yuval Lifshitz <ylifshit@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_process.cc [hunk 1]
```diff
@@ -401,16 +401,18 @@ int process_request(rgw::sal::Store* const store,
...
-    if (s->bucket) {
-      s->trace->SetAttribute(tracing::rgw::BUCKET_NAME, s->bucket->get_name());
-    }
-    if (s->object) {
-      s->trace->SetAttribute(tracing::rgw::OBJECT_NAME, s->object->get_name());
+  if (op) {
+    if (s->trace) {
+      s->trace->SetAttribute(tracing::rgw::RETURN, op->get_ret());
+      if (s->user) {
+        s->trace->SetAttribute(tracing::rgw::USER_ID, s->user->get_id().id);
```
[the remark this answers is not in the document]
Comment (yuvalif) [reply src/rgw/rgw_process.cc]: fixed. also found an issue with the sal interfaces. fixed in a separate commit in this PR

**Review by zenomri — APPROVED** [review]
Looks good to me. Thanks for the fixes.

**Comment by yuvalif** [comment]
jenkins test api

File: src/rgw/rgw_sal.h [hunk 2]
```diff
@@ -1080,7 +1080,7 @@ class Object {
     /** Check to see if the given object pointer is uninitialized */
     static bool empty(Object* o) { return (!o || o->empty()); }
     /** Check to see if the given object unique pointer is uninitialized */
-    static bool empty(std::unique_ptr<Object> o) { return (!o || o->empty()); }
+    static bool empty(std::unique_ptr<Object>& o) { return (!o || o->empty()); }
```
Comment (cbodley) [inline src/rgw/rgw_sal.h:1083]: pass by const reference please

**Review by cbodley — APPROVED** [review]

**Comment by yuvalif** [comment]
@cbodley do i need full qa on this one?

**Comment by yuvalif** [comment]
> up to you @yuvalif - do you think there's any risk of regression?

will make sure it builds in shaman (due to the changes in the "empty()" functions)

**Comment by yuvalif** [comment]
> > up to you @yuvalif - do you think there's any risk of regression?
> 
> will make sure it builds in shaman (due to the changes in the "empty()" functions)

build on shaman (except for crimson): https://shaman.ceph.com/builds/ceph/wip-yuval-fix-lua-post-ctx/e7f6962847540d747e506961473d4ebdf6d4a5d8/

**Comment by yuvalif** [comment]
jenkins test api

**Comment by yuvalif** [comment]
jenkins test api

**Comment by yuvalif** [comment]
@cbodley can we override the "API tests"?
they keep on failing for this PR - not sure why

**Comment by cbodley** [comment]
> @cbodley can we override the "API tests"?
> they keep on failing for this PR - not sure why

@epuertat any idea what's broken here? are these failures being tracked somewhere?

**Comment by yuvalif** [comment]
jenkins test api

**Comment by epuertat** [comment]
And I see that's the exact same error in previous runs ([this](https://jenkins.ceph.com/job/ceph-api/40519/) and [this](https://jenkins.ceph.com/job/ceph-api/40549/)) of the API tests. This might mean that this PR is somehow changing how the RGW API is consumed from the Dashboard. If that change is intended, then we should change that part too:

https://github.com/ceph/ceph/blob/b76b6ea16899861d6c266797b3c11d9f68f924d7/src/pybind/mgr/dashboard/controllers/rgw.py#L51-L79

https://github.com/ceph/ceph/blob/b76b6ea16899861d6c266797b3c11d9f68f924d7/src/pybind/mgr/dashboard/services/rgw_client.py#L430-L437

**Comment by epuertat** [comment]
In the mgr logs I see the following exceptions:
```
2022-07-14T10:11:36.641+0000 7f7496f36700  0 [dashboard INFO rgw_client] Found RGW daemon with configuration: host=172.21.2.9, port=8000, ssl=False
2022-07-14T10:11:36.657+0000 7f7496f36700  0 [dashboard ERROR rest_client] RGW REST API failed GET req status: 403
2022-07-14T10:11:36.661+0000 7f7496f36700  0 [dashboard ERROR rgw_client] RGW REST API failed request with status code 403
(b'{"Code":"InvalidAccessKeyId","RequestId":"tx00000aa822dbee70cb24c-0062cfebd8'
 b'-114d-default","HostId":"114d-default-default"}')

```
And:
```
2022-07-14T10:11:36.689+0000 7f7496f36700  0 [dashboard INFO request] [::ffff:172.21.2.9:44224] [GET] [404] [0.059s] [admin] [223.0B] /api/rgw/user
...
2022-07-14T10:11:39.557+0000 7f74c0bc9700  0 log_channel(audit) log [DBG] : from='client.20154 -' entity='client.admin' cmd=[{"prefix": "dashboard set-rgw-api-secret-key", "target": ["mon-mgr", ""]}]: dispatch
2022-07-14T10:11:39.965+0000 7f74c0bc9700  0 log_channel(audit) log [DBG] : from='client.20159 -' entity='client.admin' cmd=[{"prefix": "dashboard set-rgw-api-access-key", "target": ["mon-mgr", ""]}]: dispatch
2022-07-14T10:11:40.021+0000 7f7496f36700  0 [dashboard INFO rgw_client] Found RGW daemon with configuration: host=172.21.2.9, port=8000, ssl=False
2022-07-14T10:11:40.101+0000 7f7496f36700  0 [dashboard ERROR rest_client] RGW REST API failed GET, connection error (url=http://172.21.2.9:8000/admin/metadata/user?myself).
2022-07-14T10:11:40.109+0000 7f7496f36700  0 [dashboard ERROR rgw_client] RGW REST API cannot be reached. Please check your configuration and that the API endpoint is accessible
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 421, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 416, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.8/http/client.py", line 1348, in getresponse
    response.begin()
  File "/usr/lib/python3.8/http/client.py", line 316, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.8/http/client.py", line 285, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response
```

**Comment by cbodley** [comment]
@epuertat thanks for looking! :heart:

when i checked the log earlier, it looked like it was failing before it even started running tests. i've been seeing lots of random failures like that on other PRs too

> http.client.RemoteDisconnected: Remote end closed connection without response

interesting, i don't see why that would happen unless rgw crashed. @yuvalif can you please run this through teuthology to rule that out?

File: src/rgw/rgw_process.cc [hunk 3]
```diff
@@ -401,16 +401,18 @@ int process_request(rgw::sal::Store* const store,
...
+  if (op) {
+    if (s->trace) {
+      s->trace->SetAttribute(tracing::rgw::RETURN, op->get_ret());
+      if (rgw::sal::User::empty(s->user)) {
+        s->trace->SetAttribute(tracing::rgw::USER_ID, s->user->get_id().id);
+      }
+      if (rgw::sal::Bucket::empty(s->bucket)) {
+        s->trace->SetAttribute(tracing::rgw::BUCKET_NAME, s->bucket->get_name());
+      }
+      if (rgw::sal::Object::empty(s->object)) {
```
Comment (cbodley) [inline src/rgw/rgw_process.cc:413]: oh, oops. these should all be checking for `!empty()`. you're probably dereferencing nullptrs here

Comment (epuertat) [reply src/rgw/rgw_process.cc:413]: Great! We're just collecting the [mgr logs](https://jenkins.ceph.com/job/ceph-api/40655/artifact/build/out/) in the Jenkins runs, but if you think it's useful, please feel free to extend this to other logfiles too:

https://github.com/ceph/ceph-build/blob/0aa19992ffb5c7b38658d917dd154e73413f0d97/ceph-pr-api/config/definitions/ceph-pr-api.yml#L88

**Review by cbodley — APPROVED** [review]

**Comment by yuvalif** [comment]
teuthology results (`rgw:verify`): http://pulpito.front.sepia.ceph.com/yuvalif-2022-08-08_07:34:49-rgw:verify-wip-yuval-fix-lua-post-ctx-distro-default-smithi/
* valgrind issues
* `s3tests_boto3.functional.test_s3select.test_progress_expressions` failing
* `s3tests_boto3.functional.test_s3.test_multipart_upload_on_a_bucket_with_policy`

**Comment by yuvalif** [comment]
teuthology results: http://pulpito.front.sepia.ceph.com/yuvalif-2022-08-14_16:17:39-rgw-wip-yuval-fix-lua-post-ctx-distro-default-smithi/

* valgrind issues
* "Cannot create secret" failure
* "multisite" failure failing on: `AssertionError: failed meta checkpoint for zone=a2` see: http://qa-proxy.ceph.com/teuthology/yuvalif-2022-08-14_16:17:39-rgw-wip-yuval-fix-lua-post-ctx-distro-default-smithi/6972466/teuthology.log
* tempest failure: http://qa-proxy.ceph.com/teuthology/yuvalif-2022-08-14_16:17:39-rgw-wip-yuval-fix-lua-post-ctx-distro-default-smithi/6972440/teuthology.log
* timeout on: `s3tests_boto3.functional.test_s3.test_delete_tags_obj_public` (see: http://qa-proxy.ceph.com/teuthology/yuvalif-2022-08-14_16:17:39-rgw-wip-yuval-fix-lua-post-ctx-distro-default-smithi/6972445/teuthology.log)

**Pull request merged**